internal/clients/team5/iito.go

```diff
@@ -119,29 +137,65 @@ func (c *client) GetGiftResponses(receivedOffers shared.GiftOfferDict) shared.Gi
 // COMPULSORY, you need to implement this method
 
 func (c *client) UpdateGiftInfo(receivedResponses shared.GiftResponseDict) {
-	turn := c.gameState().Turn
-
-	for team, response := range receivedResponses { // for each ID
+	for team := range c.gameState().ClientLifeStatuses {
 		newGiftRequest := giftInfo{
-			requested: c.giftHistory[team].TheirRequest[turn].requested,
-			gifted:    shared.GiftOffer(response.AcceptedAmount),
-			reason:    shared.AcceptReason(response.Reason),
+			requested: c.giftHistory[team].TheirRequest[c.gameState().Turn].requested, // Amount THEY requested
+			offered:   c.giftHistory[team].TheirRequest[c.gameState().Turn].offered,   // Amount WE offered them
+			response:  receivedResponses[team],                                        // Amount THEY accepted and REASON
 		}
-		theirReq := map[uint]giftInfo{c.gameState().Turn: newGiftRequest}
-		c.giftHistory[team] = giftExchange{TheirRequest: theirReq}
+		c.giftHistory[team].TheirRequest[c.gameState().Turn] = newGiftRequest
 	}
-	c.Logf("[Debug] UpdateGiftInfo: %v", c.giftHistory[shared.Team3].TheirRequest[turn])
 }
 
 // ==================================== Gifting history to be made =========================================
 
+// ===================================== Has sending / recv gifts been implemented? ===============================
+
+// DecideGiftAmount is executed at the end of each turn and asks clients how much
+// they want to fulfill a gift offer they have made.
+// COMPULSORY, you need to implement this method
+func (c *client) DecideGiftAmount(toTeam shared.ClientID, giftOffer shared.Resources) shared.Resources {
+
+	if c.wealth() < 2 { // Imperial or Dying
+		giftOffer = 0 // Return nothing
+	}
+
+	// Regardless if we have less than 0.9 last turn we dont give gifts away
+	if c.ServerReadHandle.GetGameState().ClientInfo.Resources < 0.9*c.resourceHistory[c.gameState().Turn-1] {
```

Comment:
use `c.gameState()` instead of the whole `serverReadHandle`. Please replace all instances of this.